llvm/lib/Analysis/IR2VecAnalysis.cpp

```diff
@@ -0,0 +1,429 @@
+      Vec = lookupVocab("emptyTy");
+    } else if (Type->isLabelTy()) {
+      Vec = lookupVocab("labelTy");
+    } else if (Type->isTokenTy()) {
+      Vec = lookupVocab("tokenTy");
+    } else if (Type->isMetadataTy()) {
+      Vec = lookupVocab("metadataTy");
+    } else {
+      Vec = lookupVocab("unknownTy");
+    }
+    scaleVector(Vec, WT);
+    addVectors(InstVector, Vec);
+
+    for (auto &Op : I.operands()) {
+      Embedding Vec;
+      if (isa<Function>(Op)) {
+        Vec = lookupVocab("function");
+      } else if (isa<PointerType>(Op->getType())) {
+        Vec = lookupVocab("pointer");
+      } else if (isa<Constant>(Op)) {
+        Vec = lookupVocab("constant");
+      } else {
+        Vec = lookupVocab("variable");
+      }
+      scaleVector(Vec, WA);
+      addVectors(InstVector, Vec);
+    }
+    InstVecMap[&I] = InstVector;
+    addVectors(BBVector, InstVector);
+  }
+  return BBVector;
+}
+} // namespace
+
+// ==----------------------------------------------------------------------===//
+// IR2VecVocabResult and IR2VecVocabAnalysis
+//===----------------------------------------------------------------------===//
+
+IR2VecVocabResult::IR2VecVocabResult(ir2vec::Vocab &&Vocabulary, unsigned Dim)
+    : Vocabulary(std::move(Vocabulary)), Valid(true), DIM(Dim) {}
+
+const ir2vec::Vocab &IR2VecVocabResult::getVocabulary() const {
+  assert(Valid);
+  return Vocabulary;
+}
+
+// For now, assume vocabulary is stable unless explicitly invalidated.
+bool IR2VecVocabResult::invalidate(Module &M, const PreservedAnalyses &PA,
+                                   ModuleAnalysisManager::Invalidator &Inv) {
+  auto PAC = PA.getChecker<IR2VecVocabAnalysis>();
+  return !(PAC.preservedWhenStateless());
+}
+
+// FIXME: Make this optional. We can avoid file reads
+// by auto-generating the vocabulary during the build time.
+Error IR2VecVocabAnalysis::readVocabulary() {
+  auto BufOrError = MemoryBuffer::getFileOrSTDIN(VocabFile, /*IsText=*/true);
+  if (!BufOrError) {
+    return createFileError(VocabFile, BufOrError.getError());
+  }
+  auto Content = BufOrError.get()->getBuffer();
+  json::Path::Root Path("");
+  Expected<json::Value> ParsedVocabValue = json::parse(Content);
+  if (!ParsedVocabValue)
+    return ParsedVocabValue.takeError();
+
+  bool Res = json::fromJSON(*ParsedVocabValue, Vocabulary, Path);
+  if (!Res) {
+    return createStringError(errc::illegal_byte_sequence,
+                             "Unable to parse the vocabulary");
+  }
+  assert(Vocabulary.size() > 0 && "Vocabulary is empty");
+
+  unsigned Dim = Vocabulary.begin()->second.size();
+  assert(Dim > 0 && "Dimension of vocabulary is zero");
+  assert(std::all_of(Vocabulary.begin(), Vocabulary.end(),
+                     [Dim](const std::pair<StringRef, Embedding> &Entry) {
+                       return Entry.second.size() == Dim;
+                     }) &&
+         "All vectors in the vocabulary are not of the same dimension");
+  this->DIM = Dim;
```

Comment:
this reinforces my belief we don't need `DIM` - just have the accessor fetch it from the `Vocabulary`. Easier to maintain - it's not some state that needs to be kept in sync with anything.

Reply:
Refactored the code and removed the variable. However retained it in `Embeddings` by marking it as `const`. I thought this is better rather than querying `vocab.begin()->second.size()` each time. Please let me know if it is okay.